### PR TITLE
Fix for incorrect error message in MergeableStoreViaGetPut

### PR DESCRIPTION
**Pull request by pankajroark**
MergeableStoreViaGetPut has a bug when using FutureCollector.bestEffort. When any of the multiget requests fail they are ignored (which is correct) and afterwards all the dropped keys are mapped using missingFn which typically results in mapping to MissingValueException. The end result is that the root cause of failures of get requests is hidden and replaced with these MissingValueExceptions. This confuses users, making them think that the underlying store (such as Memcache) has lost data, whereas many times it's just timeouts. We should preserve the original exception for the get request failure, and this is what this PR is about.

But to achieve this behavior FutureCollector interface isn't sufficient, we need to get the successful results as well as the unsuccessful ones. Plus generality of using FutureCollector doesn't seem useful here. There are only two FutureCollectors currently, bestEffort and default. The default FutureCollector fails the entire multiget request if even one of the requests fails. I can't imagine any reasonable case where we'd want that behavior. So it seems best to me to remove the option of FutureCollector completely as I have proposed here. This would ultimately chain up to removing FutureCollector from MergeableStore.fromStore.

Before undertaking that api breaking change I wanted to get some feedback. Is there a valid use case for FutureCollector.default? If so any recommendations on how to organize the change. I have some ideas on how to achieve this if needed but none clean. e.g. we could fork requirements into separate functions at some level.

If the current PR looks to be on the right track I plan to remove FutureCollector for this code path and add tests for collectWithFailure.

### CODE REVIEW

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (johnynek) [inline storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala:34]: 1. the latest algebird has `Semigroup.maybePlus` this does this, but it always returns `V` not `Option[V]` which I think is correct (you can just box in `Some` as needed).

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (johnynek) [inline storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: can we put spaces in types: `fs: Map...`

Comment (johnynek) [inline storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: I'm not crazy about using `Seq`. I'd rather use `List` or `Vector` so the contracts are clearer.

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (johnynek) [inline storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: what about `AtomicReferenceArray[(K, Either[Throwable, V])]` here instead. Seems easier to follow

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (johnynek) [reply storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: If this is to be private, it can even be `Array`.

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (johnynek) [inline storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: doesn't this need to be an `AtomicReference`? How is it safe to modify this var from many threads?

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (johnynek) [inline storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: failures has to be a future right? You are returning a copy here, but the futures are not done yet. If you could have a `Map` you could see what keys have failed, and you can't can you? I think this has to be `Future[Map[K, ..]]`. I am not sure `Future[Nothing]` is a better type that `Throwable` though.

**Review by johnynek — CHANGES_REQUESTED** [review]

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (pankajroark) [reply storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala:34]: Makes sense. Don't want to merge latest algebird in this change. Is it ok to simplify this separate review where I can merge algebird. I've left a todo for now.

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (pankajroark) [reply storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: fixed

Comment (pankajroark) [reply storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: Replaced with Array.

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (pankajroark) [reply storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: Yes, that was totally wrong. I've fixed it now and added tests as well.

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (pankajroark) [reply storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: Done

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (johnynek) [inline storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: can we not use `todo` and instead file issues. We don't always close issues, but we approximately never address todos.

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (johnynek) [inline storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: looking at this again it seems like rather than `Either[Throwable, V]` we could just use `Try[V]` which is basically the same.

Then this method signature could be:
`def collectWithFailures[K, V](fs: Map[K, Future[V]]): Future[Map[K, Try[V]]]` which actually seems useful enough to have even in twitter util.

In any case, the first thing you do when you get the result is call `.toMap` on the array, so returning the `Map` seems to make sense.

Also, both of these futures are satisfied at the same time, right? So, even if you don't merge the data structures, you could do: `Future[(Map[K, V], Map[K, Throwable])]` so it is clear that all the data comes in at one time.

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (johnynek) [inline storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: if you are going to geek out on while loops above, doing the same here is probably worth it:
```scala
val mapIt = fs.iterator
var i = 0
while(mapIt.hasNext) {
  val kv = mapIt.next
  fv._2.respond {
    case Return(v) =>...
  }
  i += 0
}
```

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (pankajroark) [reply storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: The idea of Either[Throwable, V] to Try[V] is great, indeed Try is basically the same in this context and simpler.

Combining successful and failed results will be inefficient since we need to immediately do multiput for the successful ones and so need them separately. By passing them both back in a single future we have to split them up again immediately after. I do agree that returning a single future conveys the behavior more correctly. It feels like a question of performance and code simplicity vs indication of correct intent. I've implemented the return with single future PTAL.

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (pankajroark) [reply storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: Implemented this, it does add a bit of complexity. Also does avoid one wrapping/unwraping for each item in the map.

File: storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala
Comment (pankajroark) [reply storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala]: Sounds good. Not sure if filing a jira before the change is merged would be a good idea. Will file the jira after this change merges. This is a bit of problem with filing jiras in such cases, it is something that would need to be kept track separately.

**Comment by pankajroark** [comment]
I've removed couple of intermediate map creations. In this case I think optimizing is worthwhile because I've seen this function show up in YourKit profiles as taking noticeable cpu. Plus the extra maps will create garbage.

**Comment by johnynek** [comment]
:+1: when tests are green. Nice result. Good work removing more intermediate Map creations.

**Comment by pankajroark** [comment]
So, the binary compatibility tests are failing because because the signature of the function has changed. This is another thing I wanted feedback on, that it is ok to make this binary incompatible change. Also, how do I make the tests pass in case of binary incompatible changes, by incrementing version?

**Comment by sritchie-stripe** [comment]
Hey @pankajroark,

I'd say it's fine for now since we have so damned many binary incompatible changes stacked up. To make the tests pass, copy the suggested line from the mima check in the build into here:

https://github.com/twitter/storehaus/blob/develop/build.sbt#L31

Once that's done we'll merge!

**Comment by johnynek** [comment]
can't you implement the old function still? Do we need to add the binary incompat?

To me, each binary (and source) incompatible change puts a cost out on users, and slows adoption (people get stuck on old versions rather than pay the tax to move up).

If it it easy (or only slightly a pain), we should keep source and binary compatibility.

In this case, it is easy to add a second method where the user passes a function in, right? you are basically inlining the old default on line 75, right?

**Comment by pankajroark** [comment]
MissingValueException there will actually never hit, I kept it there just
for safety. Letting users supply that would be misleading. Also we don't
use the future collector. FutureCollector doesn't allow capturing
individual failures.

On Thu, Dec 15, 2016 at 1:13 PM P. Oscar Boykin <notifications@github.com>
wrote:

can't you implement the old function still? Do we need to add the binary
incompat?

To me, each binary (and source) incompatible change puts a cost out on
users, and slows adoption (people get stuck on old versions rather than pay
the tax to move up).

If it it easy (or only slightly a pain), we should keep source and binary
compatibility.

In this case, it is easy to add a second method where the user passes a
function in, right? you are basically inlining the old default on line 75,
right?

—
You are receiving this because you were mentioned.


Reply to this email directly, view it on GitHub
<https://github.com/twitter/storehaus/pull/334#issuecomment-267445579>, or mute
the thread

**Comment by johnynek** [comment]
okay, if you are fairly sure that the old API does not make sense anymore, I guess. But adding a @deprecated and calling the new API can help users other than yourself.

For instance, we get stuck on all kinds of old versions at Stripe over little issues like this. Each one may not be a big deal, but when you try to bump summingbird and you have to bump util, finagle, bijection, algebird, scalding and storehaus, and each has a few incompatibilities, it turns into a really big job.

We can add this method to this list:
https://github.com/twitter/storehaus/blob/develop/build.sbt#L28

to get the tests green (keeping in mind we should do a version bump next and remove all that list).

**Comment by pankajroark** [comment]
Thanks for the guidance, will do that.

**Review by johnynek — APPROVED** [review]

**Comment by johnynek** [comment]
👍

**Pull request merged**
